src/js/fable-core/Array2.fs

```diff
@@ -1,16 +1,732 @@
+    source |> mapIndexed (fun i x -> i, x) <| cons;
+
+let private typedArrayConcatImpl (cons: ArrayCons) (arrays: 'T[][]): 'T[] =
+    let mutable totalLength = 0
+    for arr in arrays do
+        totalLength <- totalLength + arr.Length
+
+    let result = cons.Create totalLength
+    let mutable offset = 0;
+    for arr in arrays do
+        typedArraySetImpl result arr <| float(offset)
+        offset <- offset + arr.Length
+    result
+
+let private concatImpl (cons: ArrayCons) (arrays: 'T[][]): 'T[] =
+    if arrays.Length > 0 then
+        if isTypedArray arrays.[0] then
+            typedArrayConcatImpl cons arrays
+        else
+            dynamicArrayConcatImpl arrays
+    else
+        cons.Create 0
+
+let concat (arrays: 'T[] seq) (cons: ArrayCons): 'T[] =
+    arrays
+    |> Seq.toArray
+    |> concatImpl cons
+
+let collect (mapping: 'T -> 'U[]) (array: 'T[]) (cons: ArrayCons): 'U[] =
+    map mapping array cons
+    |> concatImpl cons
+
+let countBy (projection: 'T->'Key) (array: 'T[]) (cons: ArrayCons) =
+    let dict = System.Collections.Generic.Dictionary<'Key, int>()
+
+    for value in array do
+        let key = projection value
+        let mutable prev = Unchecked.defaultof<_>
+        if dict.TryGetValue(key, &prev) then dict.[key] <- prev + 1 else dict.[key] <- 1
+
+    let res = cons.Create dict.Count
+    let mutable i = 0
+    for group in dict do
+        res.[i] <- group.Key, group.Value
+        i <- i + 1
+    res
+
+let distinctBy projection (array:'T[]) (cons: ArrayCons) =
+    let temp = cons.Create array.Length
+    let mutable i = 0
+
+    let hashSet = System.Collections.Generic.HashSet<'T>()
+    for v in array do
+        if hashSet.Add(projection v) then
+            temp.[i] <- v
+            i <- i + 1
+
+    getSubArray temp 0 i
+
+let distinct (array: 'T[]) (cons: ArrayCons) = distinctBy id array cons
+
+let where predicate (array: _[]) = filter predicate array
+
+let except (itemsToExclude: seq<_>) (array:_[]) =
+    if array.Length = 0 then
+        array
+    else
+        let cached = System.Collections.Generic.HashSet(itemsToExclude)
+        array |> filter cached.Add
+
+let groupBy (projection: 'T->'Key) (array: 'T[]) (cons: ArrayCons) =
+    let dict = System.Collections.Generic.Dictionary<'Key, 'T[]>()
+
+    // Build the groupings
+    for i = 0 to (array.Length - 1) do
+        let v = array.[i]
+        let key = projection v
+        let mutable prev = Unchecked.defaultof<_>
+        if dict.TryGetValue(key, &prev) then
+            push prev v |> ignore
+        else
+            let prev = cons.From [|v|]
+            dict.[key] <- prev
+
+    // Return the array-of-arrays.
+    let result = cons.Create dict.Count
+    let mutable i = 0
+    for group in dict do
+        result.[i] <- group.Key, group.Value
+        i <- i + 1
+
+    result
+
+let ofArray (arr: 'T[]) = arr
+
+let empty<'T> : 'T[] = [||] // TODO: Shoud it be a ctor function with cons.Create 0?
```

Comment:
As this is an empty array, I guess it doesn't matter if it's empty or not. You can also use `Array.empty`.

Reply:
So we don't mind if it's `[]` or, say, `new Int8Array()` here? Sounds confusing to me. Am I missing something?

Reply:
Sorry, I noticed a typo in my comment above, I just fixed it. In theory, using `[]` or `new Int8Array(0)` should be the same, because Fable is prepared to accept both dynamic and typed arrays. But it's true this can affect JS optimizations expecting monomorphic functions so it may be better to use `cons.Create(0)` instead.